docs/Style.thy

```diff
@@ -0,0 +1,260 @@
+style see Gerwin's style guide (parts 1 and 2).
+
+  * https://proofcraft.org/blog/isabelle-style.html
+  * https://proofcraft.org/blog/isabelle-style-part2.html
+\<close>
+
+section \<open>General Principles\<close>
+
+text \<open>
+  Goals:
+  * Readability and consistency:
+      reduce cognitive overhead.
+
+  * Make good use of horizontal space:
+      most screens are wide, not high
+
+  * Semantic indentation:
+      where possible group things according to meaning. But remain syntactically consistent:
+      a sub-term should not go more left than its parent term.
+
+  * Support tools like diff and grep:
+      diff tries to find the "function" each chunk belongs to by indentation. If we remain
+      compatible, diffs are much easier to read. Pull requests mean we read a lot of diffs.
+      grep and other regexp search is not needed as much as it used to, but it's still often
+      useful to be able to find constants and lemmas by regexp search outside of Isabelle
+
+  * Don't fight existing auto-indent where possible:
+      there can be exceptions, but each time we go against defaults, it will be annoying to follow.
+
+  * Compatibility with upstream style where it exists:
+      there is not much consistent Isabelle distribution style, but where there is, we should follow
+      it, so that others can read our code and we can contribute upstream
+\<close>
+
+text \<open>
+The main rules:
+  * Indent your proofs.
+  * Follow naming conventions.
+  * Be aware that others have to read what you write.
+
+To not drive proof maintainers insane:
+  * Do not use auto except at the end of a proof.
+  * Never check in proofs containing back.
+  * Instantiate bound variables in preference to leaving schematics in a subgoal
+    (i.e. use erule_tac ... allE in preference to erule allE)
+  * Explicitly name variables that the proof text refers to (e.g. with rename_tac)
+  * Don't mix object and meta logic in a lemma statement.
+\<close>
+
+section \<open>Indentation\<close>
+
+text \<open>
+Isabelle code is much easier to maintain when indented consistently. In apply style proofs we
+indent by 2 spaces, and add an additional space for every additional subgoal. For example, the
+rules iffI and conjI add a new subgoal, and fast removes a subgoal. The idea is that, when
+something breaks, the indentation tells you whether a tactic used to solve a subgoal or produced
+new ones.
+\<close>
+
+lemma disj_conj_distribR:
+  "(A \<and> B \<or> C) = ((A \<or> C) \<and> (B \<or> C))"
+  apply (rule iffI)
+   apply (rule conjI)
+    apply fast
+   apply fast
+  apply auto
+  done
+
+text \<open>
+The statement of Hoare triple lemmas are generally indented in the following way.
+\<close>
+
+lemma my_hoare_triple_lemma:
+  "\<lbrace> precondition_one and precondition_two and
+     precondition three \<rbrace>
+   my_function param_a param_b
+   \<lbrace> post_condition \<rbrace>"
```

Comment:
No real argument, but a mild preference for no spaces, if only because it's the way I'm accustomed to the syntax translations printing things back to me.

Reply:
Also mild preference for no spaces.